modules/luci-base/root/etc/init.d/ucitrack

```diff
@@ -0,0 +1,57 @@
+#!/bin/sh /etc/rc.common
+
+START=80
+USE_PROCD=1
+
+register_init() {
+	local config="$1"
+	local init="$2"
+	shift; shift
+
+	if [ -x "$init" ] && "$init" enabled && ! grep -sqE 'USE_PROCD=.' "$init"; then
+		logger -t "ucitrack" "Setting up /etc/config/$config reload trigger for non-procd $init"
+		procd_add_config_trigger "config.change" "$config" "$init" "$@"
+	fi
+}
+
+register_trigger() {
+	local sid="$1"
+	local config init exec affects affected
+
+	config_get config "$sid" TYPE
+	config_get init "$sid" init
+	config_get exec "$sid" exec
+	config_get affects "$sid" affects
+
+	if [ -n "$init" ]; then
+		register_init "$config" "/etc/init.d/$init" "reload"
+	fi
+
+	if [ -n "$exec" ]; then
+		case "$exec" in
+			/etc/init.d/*)
+				set -- $exec
+				register_init "$config" "$@"
+			;;
+			*)
+				logger -t "ucitrack" "Setting up non-init /etc/config/$config reload handler: $exec"
+				procd_add_config_trigger "config.change" "$config" "$exec"
+			;;
+		esac
+	fi
+
+	for affected in $affects; do
```

Comment:
I think this code reverses `$config` and `$affected`, as well as ignores whether the `$affected` init uses `procd`.

e.g. it would be simpler and more correct to do something like:
```
	for affected in $affects; do
		register_init "$config" "/etc/init.d/$affected" "reload"
	done
```
but even that ignores the case of `$affected` having an `exec` entry in `ucitrack`.

Perhaps restructure rewrite `register_init()` to be `register_init_or_exec()`, and then use the latter to iterate over the combined list of `$config` and `$affects`.

Hopefully I didn't miss something obvious. :-)

Reply:
The logic is intentionally reversed; procd already takes care of reloading services whose own configs changed and LuCI will emit config.change notifications for those. This code here sets up the reverse dependencies procd does not know about, meaning we will reload *our* service if any of the potentially affected configs changed.

We also assume that affected entries are either procd enabled or covered by ucitrack as well (end thus have a dummy service entry reacting on reload signals).

Reply:
I'm agreeing with what you say, just thinking the implementation doesn't line up. Let's take a concrete example of `/etc/config/ucitrack`:
```
config firewall
        option init 'firewall'
        list affects 'sqm'
config sqm
        option init 'sqm'
```
This means that `firewall` config changes should cause `sqm` service to reload (and rightly so) in the old "luci-reload" world.

In the new world of "procd + luci rollback", `/etc/init.d/{firewall,sqm}` are each procd-enabled and add a trigger to self-reload on changes to their own config. So far so good. But why then does `/etc/init.d/ucitrack` handle the `list affects 'sqm'` by issuing the procd command:
```
        procd_add_config_trigger "config.change" "sqm" \
            ubus call service event \
            '{"type":"config.change","data":{"package":"firewall"}}'
```
The `firewall` service *should not* be reloaded on a `sqm` config change. This is opposite to the `/etc/config/ucitrack` config above.

Help me understand why you think this opposite behaviour is correct. I must be missing something, or I'm up too late.

As an aside, my own `sqm-scripts` fork also includes a `procd_add_reload_trigger "firewall"` precisely to remove any `ucitrack` dependency. Isn't that the right approach? That's why I thought it reasonable to check if an "affected" init script is not procd-enabled, and only then issue the correct "cross-service" config trigger.

Reply:
> The firewall service should not be reloaded on a sqm config change. This is opposite to the /etc/config/ucitrack config above.

The firewall should get reloaded. The old luci-reload (and luci.model.uci.apply()) implementation issued an init reload for each affected package including itself. So if any of sqm or firewall got changed, both sqm and firewall got reloaded. The above takes care of doing that in the reverse direction by broadcasting a change event and letting procd sort out the details.

> As an aside, my own sqm-scripts fork also includes a procd_add_reload_trigger "firewall" precisely to remove any ucitrack dependency. Isn't that the right approach?

It is.

> That's why I thought it reasonable to check if an "affected" init script is not procd-enabled, and only then issue the correct "cross-service" config trigger.

If an init script is procd-enabled, the extraneous config.change event is harmless. Procd will a) caolesce exevents and b) no-op the change event if the calculated desired runtime change is identical to the effective runtime state (it compares command lines and keeps checksums of all related configurations).

Reply:
Please also keep in mind that this is a transitional mechanism. I plan to fade out ucitrack entirely within the next few months. No package should rely on it anymore. The sole reason I did that is to keep package not converted to procd somewhat working yet.

Reply:
Hmmm. I had set aside the problems I saw on the expectation that `ucitrack` would be quickly phased out. However, it's now been more than 2 years since this transitional mechanism was to be retired, and the same issues persist. What are the remaining obstacles to fade out `ucitrack` entirely as planned?